pyteal/ast/abi/address.py

```diff
@@ -1,33 +1,120 @@
+from collections.abc import Sequence as CollectionSequence
+
+from pyteal.errors import TealInputError
+
+from pyteal.ast.bytes import Bytes
+from pyteal.ast.addr import Addr
+from pyteal.ast.abi.type import ComputedValue, BaseType
 from pyteal.ast.abi.array_static import StaticArray, StaticArrayTypeSpec
-from pyteal.ast.abi.uint import ByteTypeSpec
+from pyteal.ast.abi.uint import ByteTypeSpec, Byte
 from pyteal.ast.expr import Expr
 
-ADDRESS_LENGTH = 32
+
+class AddressLength(Enum):
+    String = 58
+    Bytes = 32
+
+
+AddressLength.__module__ = "pyteal"
 
 
 class AddressTypeSpec(StaticArrayTypeSpec):
     def __init__(self) -> None:
-        super().__init__(ByteTypeSpec(), ADDRESS_LENGTH)
+        super().__init__(ByteTypeSpec(), AddressLength.Bytes.value)
 
     def new_instance(self) -> "Address":
         return Address()
 
     def __str__(self) -> str:
         return "address"
 
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, AddressTypeSpec)
+
 
 AddressTypeSpec.__module__ = "pyteal"
 
 
-class Address(StaticArray):
+class Address(StaticArray[Byte, Literal[32]]):
     def __init__(self) -> None:
-        super().__init__(AddressTypeSpec(), ADDRESS_LENGTH)
+        super().__init__(AddressTypeSpec())
 
     def type_spec(self) -> AddressTypeSpec:
         return AddressTypeSpec()
 
     def get(self) -> Expr:
         return self.stored_value.load()
 
+    def set(
+        self,
+        value: Union[
+            Sequence[Byte],
+            StaticArray[Byte, Literal[32]],
+            ComputedValue[StaticArray[Byte, Literal[32]]],
+            "Address",
+            str,
+            bytes,
+            Expr,
+        ],
+    ):
+
+        match value:
+            case ComputedValue():
+                pts = value.produced_type_spec()
+                if pts == AddressTypeSpec() or pts == StaticArrayTypeSpec(
+                    ByteTypeSpec(), AddressLength.Bytes.value
+                ):
+                    return value.store_into(self)
+
+                raise TealInputError(
+                    f"Got ComputedValue with type spec {pts}, expected AddressTypeSpec or StaticArray[Byte, Literal[32]]"
+                )
+            case BaseType():
+                if (
+                    value.type_spec() == AddressTypeSpec()
+                    or value.type_spec()
+                    == StaticArrayTypeSpec(ByteTypeSpec(), AddressLength.Bytes.value)
+                ):
+                    return self.stored_value.store(value.stored_value.load())
+
+                raise TealInputError(
+                    f"Got {value} with type spec {value.type_spec()}, expected AddressTypeSpec"
+                )
+            case str():
+                if len(value) == AddressLength.String.value:
+                    return self.stored_value.store(Addr(value))
+                raise TealInputError(
+                    f"Got string with length {len(value)}, expected {AddressLength.String.value}"
```

Comment:
actually, if you noticed https://github.com/algorand/pyteal/blob/master/pyteal/ast/addr.py#L21:
the implementation https://github.com/algorand/pyteal/blob/master/pyteal/types.py#L51 checks whether the string length is *exactly* 58, wonder if this would help make implementation lighter.